File: mpp-proiect-java-culbec/MotorcycleContest/Networking/src/main/java/Protocol/ProtocolUtils.java
package Protocol;

import Model.Participant;
import Model.Race;
import Model.User;

import java.util.stream.StreamSupport;

public class ProtocolUtils {
    /**
     * Creates a request for logging in.
     * @param username The username of the user.
     * @param password The password of the user.
     * @return A request for logging in.
     */
    public static NetworkProtos.Request createLoginRequest(String username, String password) {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.LOGIN)
                .setUsername(username)
                .setPassword(password)
                .build();
    }

    /**
     * Creates a request for logging out.
     * @param user The user to log out.
     * @return A request for logging out.
     */
    public static NetworkProtos.Request createLogoutRequest(User user) {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.LOGOUT)
                .setUser(NetworkProtos.User.newBuilder()
                        .setId(user.getId())
                        .setFirstName(user.getFirstName())
                        .setLastName(user.getLastName())
                        .setUsername(user.getUsername())
                        .build())
                .build();
    }

    /**
     * Creates a request for adding a participant.
     * @param firstName The first name of the participant.
     * @param lastName The last name of the participant.
     * @param team The team of the participant.
     * @param engineCapacity The engine capacity of the participant's motorcycle.
     * @return A request for adding a participant.
     */
    public static NetworkProtos.Request createAddParticipantRequest(String firstName, String lastName, String team, int engineCapacity) {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.ADD_PARTICIPANT)
                .setParticipant(NetworkProtos.Participant.newBuilder()
                        .setFirstName(firstName)
                        .setLastName(lastName)
                        .setTeam(team)
                        .setEngineCapacity(engineCapacity)
                        .build())
                .build();
    }

    /**
     * Creates a request for finding participants by team.
     * @param team The team to search for.
     * @return A request for finding participants by team.
     */
    public static NetworkProtos.Request createFindParticipantsByTeamRequest(String team) {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.FIND_PARTICIPANTS_BY_TEAM)
                .setTeam(team)
                .build();
    }

    /**
     * Creates a request to find all the saved races.
     * @return A request to find all the saved races.
     */
    public static NetworkProtos.Request createFindAllRacesRequest() {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.FIND_RACES)
                .build();
    }

    /**
     * Creates a request to find all the engine capacities of the saved races.
     * @return A request to find all the engine capacities of the saved races.
     */
    public static NetworkProtos.Request createFindAllRaceEngineCapacitiesRequest() {
        return NetworkProtos.Request.newBuilder()
                .setRequestType(NetworkProtos.Request.type.FIND_ENGINE_CAPACITIES)
                .build();
    }

    // OK response.
    public static NetworkProtos.Response OKResponse = NetworkProtos.Response.newBuilder()
            .setResponseType(NetworkProtos.Response.type.OK)
            .build();

    /**
     * Creates an error response.
     * @param message The error message.
     * @return An error response.
     */
    public static NetworkProtos.Response createErrorResponse(String message) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.ERROR)
                .setErrorMessage(message)
                .build();
    }

    /**
     * Creates a connection closed response.
     * @return The connection closed response.
     */
    public static NetworkProtos.Response createConnectionClosedResponse() {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.CONNECTION_CLOSED)
                .build();
    }

    /**
     * Creates a response for logging in.
     * @param user The user that logged in.
     * @return A response for logging in.
     */
    public static NetworkProtos.Response createLoginResponse(User user) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .setUser(NetworkProtos.User.newBuilder()
                        .setId(user.getId())
                        .setFirstName(user.getFirstName())
                        .setLastName(user.getLastName())
                        .setUsername(user.getUsername())
                        .build())
                .build();
    }

    /**
     * Creates a response for logging out.
     * @return A response for logging out.
     */
    public static NetworkProtos.Response createLogoutResponse() {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .build();
    }

    /**
     * Creates a response for adding a participant.
     * @param participant The participant that was added.
     * @return A response for adding a participant.
     */
    public static NetworkProtos.Response createAddParticipantResponse(NetworkProtos.Participant participant) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .setParticipant(participant)
                .build();
    }

    /**
     * Creates a response for finding participants by team.
     * @param participant The participant that was found.
     * @return A response for finding participants by team.
     */
    public static NetworkProtos.Response createParticipantAddedResponse(Participant participant) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.PARTICIPANT_ADDED)
                .setParticipant(NetworkProtos.Participant.newBuilder()
                        .setFirstName(participant.getFirstName())
                        .setLastName(participant.getLastName())
                        .setTeam(participant.getTeam())
                        .setEngineCapacity(participant.getEngineCapacity())
                        .build())
                .build();
    }

    /**
     * Creates a response for finding participants by team.
     * @param participants The participants that were found.
     * @return A response for finding participants by team.
     */
    public static NetworkProtos.Response createFindParticipantsByTeamResponse(Iterable<Participant> participants) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .addAllParticipants(StreamSupport.stream(participants.spliterator(), false)
                        .map(participant -> NetworkProtos.Participant.newBuilder()
                                .setId(NetworkProtos.UUID.newBuilder()
                                        .setValue(participant.getId().toString()).build())
                                .setFirstName(participant.getFirstName())
                                .setLastName(participant.getLastName())
                                .setTeam(participant.getTeam())
                                .setEngineCapacity(participant.getEngineCapacity())
                                .build())
                        .toList())
                .build();
    }

    /**
     * Creates a response for finding all saved races.
     * @param races Races saved.
     * @return A response for finding all saved races.
     */
    public static NetworkProtos.Response createFindAllRacesResponse(Iterable<Race> races) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .addAllRaces(StreamSupport.stream(races.spliterator(), false)
                        .map(race -> NetworkProtos.Race.newBuilder()
                                .setId(race.getId())
                                .setEngineCapacity(race.getEngineCapacity())
                                .setNoParticipants(race.getNoParticipants())
                                .build())
                        .toList())
                .build();
    }

    /**
     * Creates a response for finding all engine capacities of the saved races.
     * @param engineCapacities Engine capacities of the saved races.
     * @return A response for finding all engine capacities of the saved races.
     */
    public static NetworkProtos.Response createFindAllRaceEngineCapacitiesResponse(Iterable<Integer> engineCapacities) {
        return NetworkProtos.Response.newBuilder()
                .setResponseType(NetworkProtos.Response.type.OK)
                .addAllEngineCapacities(engineCapacities)
                .build();
    }

}

File: mpp-proiect-csharp-culbec/Network/Protocol/ProtocolUtils.cs
using NetworkProtos.Protocol;

namespace Network.Protocol;

public static class ProtocolUtils
{
    /// <summary>
    /// Creates a request for the login operation.
    /// </summary>
    /// <param name="username">Username of the user that tries to login.</param>
    /// <param name="password">Password of the user that tries to login.</param>
    /// <returns>A Protobuf Request for the login operation.</returns>
    public static Request CreateLoginRequest(string username, string password)
    {
        var request = new Request
        {
            RequestType = Request.Types.type.Login,
            Username = username,
            Password = password
        };

        return request;
    }

    /// <summary>
    /// Creates a request for the logout operation.
    /// </summary>
    /// <param name="user">User that wants to logout.</param>
    /// <returns>A Protobuf Request for the logout operation.</returns>
    public static Request CreateLogoutRequest(Model.User user)
    {
        var request = new Request
        {
            RequestType = Request.Types.type.Logout,
            User = new User
            {
                Id = user.Id,
                FirstName = user.FirstName,
                LastName = user.LastName,
                Username = user.Username
            }
        };

        return request;
    }

    /// <summary>
    /// Creates a request for the add participant operation.
    /// </summary>
    /// <param name="firstname">First name of the participant.</param>
    /// <param name="lastname">Last name of the participant.</param>
    /// <param name="team">Team of the participant.</param>
    /// <param name="engineCapacity">Engine capacity that the participant enrolled with.</param>
    /// <returns>A Protobuf Request for the add participant operation.</returns>
    public static Request CreateAddParticipantRequest(string firstname, string lastname, string team, int engineCapacity)
    {
        var request = new Request
        {
            RequestType = Request.Types.type.AddParticipant,
            Participant = new Participant
            {
                FirstName = firstname,
                LastName = lastname,
                Team = team,
                EngineCapacity = engineCapacity
            }
        };

        return request;
    }

    /// <summary>
    /// Creates a request for the find all participants of a given team operation.
    /// </summary>
    /// <param name="team">Team of the participants.</param>
    /// <returns>A Protobuf Request to find all participant of a given team operation.</returns>
    public static Request CreateFindParticipantsByTeamRequest(string team)
    {
        var request = new Request
        {
            RequestType = Request.Types.type.FindParticipantsByTeam,
            Team = team
        };

        return request;
    }

    /// <summary>
    /// Creates a request for the find all races operation.
    /// </summary>
    /// <returns>A Protobuf Request to find all races operation.</returns>
    public static Request CreateFindAllRacesRequest()
    {
        var request = new Request
        {
            RequestType = Request.Types.type.FindRaces
        };

        return request;
    }

    /// <summary>
    /// Creates a request for the find all engine capacities operation.
    /// </summary>
    /// <returns>A Protobuf Request to find all engine capacities operation.</returns>
    public static Request CreateFindAllRaceEngineCapacitiesRequest()
    {
        var request = new Request
        {
            RequestType = Request.Types.type.FindEngineCapacities
        };

        return request;
    }

    /// <summary>
    /// Creates a response for the login operation.
    /// </summary>
    /// <param name="user">User that logged in.</param>
    /// <returns>A Protobuf Response for the login operation.</returns>
    public static Response CreateLoginResponse(Model.User user)
    {
        var response = new Response
        {
            ResponseType = Response.Types.type.Ok,
            User = new User
            {
                Id = user.Id,
                FirstName = user.FirstName,
                LastName = user.LastName,
                Username = user.Username
            }
        };

        return response;
    }

    /// <summary>
    /// Creates a response for the logout operation.
    /// </summary>
    /// <returns>A Protobuf Response for the logout operation.</returns>
    public static Response CreateLogoutResponse()
    {
        return new Response
        {
            ResponseType = Response.Types.type.Ok
        };
    }

    /// <summary>
    /// Creates a response for the add participant operation.
    /// </summary>
    /// <param name="participant">Participant that was added.</param>
    /// <returns>A Protobuf Response for the add participant operation.</returns>
    public static Response CreateAddParticipantResponse(Participant participant)
    {
        var response = new Response
        {
            ResponseType = Response.Types.type.Ok,
            Participant = participant
        };

        return response;
    }

    /// <summary>
    /// Creates a response for the participant added event for the observers.
    /// </summary>
    /// <param name="participant">Participant that was added.</param>
    /// <returns>A Protobuf Response for the participant added event</returns>
    public static Response CreateParticipantAddedResponse(Model.Participant participant)
    {
        var response = new Response
        {
            ResponseType = Response.Types.type.ParticipantAdded,
            Participant = new Participant
            {
                Id = new UUID
                {
                    Value = participant.Id.ToString()
                },
                FirstName = participant.FirstName,
                LastName = participant.LastName,
                Team = participant.Team,
                EngineCapacity = participant.EngineCapacity
            }
        };

        return response;
    }

    /// <summary>
    /// Creates a response for the find participants by a given team operation.
    /// </summary>
    /// <param name="participants">List of the found participants of the given team.</param>
    /// <returns>A Protobuf Response for the find participants by a given team operation.</returns>
    public static Response CreateFindParticipantsByTeamResponse(IEnumerable<Model.Participant> participants)
    {
        var participantProtos = participants.Select((participant, _) => new Participant
        {
            Id = new UUID
            {
                Value = participant.Id.ToString()
            },
            FirstName = participant.FirstName,
            LastName = participant.LastName,
            Team = participant.Team,
            EngineCapacity = participant.EngineCapacity
        }).ToList();

        var response = new Response
        {
            ResponseType = Response.Types.type.Ok,
            Participants = { participantProtos }
        };

        return response;
    }

    /// <summary>
    /// Creates a response for the find all races operation.
    /// </summary>
    /// <param name="races">Races that were found.</param>
    /// <returns>A Protobuf Response for the find all races operation.</returns>
    public static Response CreateFindAllRacesResponse(IEnumerable<Model.Race> races)
    {
        var raceProtos = races.Select((race, _) => new Race
        {
            Id = race.Id,
            EngineCapacity = race.EngineCapacity,
            NoParticipants = race.NoParticipants
        }).ToList();

        var response = new Response
        {
            ResponseType = Response.Types.type.Ok,
            Races = { raceProtos }
        };
        return response;
    }

    /// <summary>
    /// Creates a response for the find all engine capacities operation.
    /// </summary>
    /// <param name="engineCapacities">Engine capacities that were found.</param>
    /// <returns>A Protobuf Response for the find all engine capacities operation.</returns>
    public static Response CreateFindAllRaceEngineCapacitiesResponse(IEnumerable<int> engineCapacities)
    {
        var response = new Response
        {
            ResponseType = Response.Types.type.Ok,
            EngineCapacities = { engineCapacities }
        };

        return response;
    }

    public static readonly Response OkResponse = new()
    {
        ResponseType = Response.Types.type.Ok
    };

    /// <summary>
    /// Creates an error response.
    /// </summary>
    /// <param name="message">Message of the error.</param>
    /// <returns>The new error response.</returns>
    public static Response CreateErrorResponse(string message)
    {
        return new Response
        {
            ResponseType = Response.Types.type.Error,
            ErrorMessage = message
        };
    }

    /// <summary>
    /// Creates a closed connection response.
    /// </summary>
    /// <returns>Closed connection response.</returns>
    public static Response CreateConnectionClosedResponse()
    {
        return new Response
        {
            ResponseType = Response.Types.type.ConnectionClosed
        };
    }
}